src/test/java/org/jenkinsci/plugins/ewm/steps/ExwsAllocateStepTest.java

```diff
@@ -0,0 +1,247 @@
+        createUpstreamJobAndRun();
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertLogContains(format("Physical path on disk was not provided in the Jenkins global config for Disk ID: '%s', within Disk Pool ID '%s'", DISK_ID_ONE, DISK_POOL_ID), upstreamRun);
+    }
+
+    @Test
+    public void physicalPathOnDiskNotRelative() throws Exception {
+        setUpDiskPool(new Disk(DISK_ID_ONE, "name", "mount", "/path"));
+        createUpstreamJobAndRun();
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertLogContains(format("Physical path on disk defined for Disk ID '%s', within Disk Pool ID '%s' must be a relative path", DISK_ID_ONE, DISK_POOL_ID), upstreamRun);
+    }
+
+    @Test
+    public void successfullyAllocateWorkspace() throws Exception {
+        Disk disk1 = new Disk(DISK_ID_ONE, "name", pathToDisk1.getPath(), "path");
+        Disk disk2 = new Disk(DISK_ID_TWO, "name", pathToDisk2.getPath(), "path");
+        setUpDiskPool(disk1, disk2);
+        createUpstreamJobAndRun();
+
+        Disk allocatedDisk = findAllocatedDisk(disk1, disk2);
+
+        j.assertBuildStatusSuccess(upstreamRun);
+        j.assertLogContains(format("Selected Disk ID '%s' from the Disk Pool ID '%s'", allocatedDisk.getDiskId(), DISK_POOL_ID), upstreamRun);
+        j.assertLogContains(format("The path on Disk is: %s/%s/%d", allocatedDisk.getPhysicalPathOnDisk(), upstreamRun.getParent().getFullDisplayName(), upstreamRun.getNumber()), upstreamRun);
+    }
+
+    /* ##### Tests for the downstream Job ###### */
+
+    @Test
+    public void downstreamJobWithInvalidUpstreamName() throws Exception {
+        String upstreamName = "random-upstream-name";
+        createDownstreamJobAndRun(upstreamName);
+
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("Can't find any upstream Jenkins job by the full name '%s'. Are you sure that this is the full project name?", upstreamName), downstreamRun);
+    }
+
+    @Test
+    public void upstreamJobHasNoStableBuild() throws Exception {
+        createUpstreamJobAndRun("random-disk-pool");
+        String upstreamName = upstreamRun.getParent().getName();
+        createDownstreamJobAndRun(upstreamName);
+
+        j.assertBuildStatus(FAILURE, upstreamRun);
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("'%s' doesn't have any stable build", upstreamName), downstreamRun);
+    }
+
+    @Test
+    public void upstreamJobHasNoActionRegistered() throws Exception {
+        String jobName = randomAlphanumeric(10);
+        createWorkflowJobAndRun(jobName, "echo 'hello world'");
+        createDownstreamJobAndRun(jobName);
+
+        j.assertBuildStatus(FAILURE, downstreamRun);
+        j.assertLogContains(format("The Jenkins job '%s' does not have registered any 'External Workspace Allocate' action. Did you run exwsAllocate step in the upstream job?", jobName), downstreamRun);
```

Comment:
Please let me know if you find future grammar mistakes, since my english has some flaws :)

Regarding the corner cases:

> .2. upstream calls exwsAllocate but fails before it registers the action.

and

> .2. upstream calls exwsAllocate but fails after it registers the action
If the build fails, no matter at which point, that build number will not be taken into consideration anyway.

But I can improve this error message for the case when the upstream build is green, but the user didn't call exwsAllocate step.